libclc/clc/lib/generic/common/clc_sign.cl

```diff
@@ -0,0 +1,38 @@
+#include <clc/clcmacro.h>
+#include <clc/internal/clc.h>
+#include <clc/relational/clc_isnan.h>
+
+#define CLC_SIGN(TYPE, F)                                                      \
+  _CLC_DEF _CLC_OVERLOAD TYPE __clc_sign(TYPE x) {                             \
+    if (__clc_isnan(x)) {                                                      \
+      return 0.0F;                                                             \
+    }                                                                          \
+    if (x > 0.0F) {                                                            \
+      return 1.0F;                                                             \
+    }                                                                          \
+    if (x < 0.0F) {                                                            \
+      return -1.0F;                                                            \
+    }                                                                          \
+    return x; /* -0.0 or +0.0 */                                               \
```

Comment:
copysign(0.0, x < 0.0)?

Reply:
The exact code I posted should be a correct implementation. The bit twiddling ideally would also form the copysign intrinsic

Separate PR from this though

Reply:
OpenCL sign returns 1.0 if x > 0 and -1.0 if x < 0. copysign also returns NaN for `copysign(NaN, ...)`, whereas sign should return 0.0.

Reply:
You swapped the operands. 0 is the magnitude in the snippet I posted

Reply:
True, my mistake. Forgetting the NaNs, though, my first sentence holds: `copysign(0.0, x < 0.0)` doesn't return `1.0` if `x > 0` nor `-1.0` if `x < 0` as OpenCL's `sign` is required to. I'm assuming you're saying `float sign(float x) { return copysign(0.0, x < 0); }` is a correct implementation of `sign`.